logsmanagement/db_api.c

```diff
@@ -0,0 +1,1096 @@
+/** @file db_api.c
+ *  @brief This is the file containing the database API
+ *
+ *  @author Dimitris Pantazis
+ */
+
+#include "../daemon/common.h"
+#include "db_api.h"
+#include <inttypes.h>
+#include <stdio.h>
+#include "circular_buffer.h"
+#include "compression.h"
+#include "helper.h"
+#include "lz4.h"
+#include "parser.h"
+
+#define MAIN_DB "main.db" /**< Primary DB with just 1 table - MAIN_COLLECTIONS_TABLE **/
+#define MAIN_COLLECTIONS_TABLE "LogCollections"
+#define BLOB_STORE_FILENAME "logs.bin"
+#define METADATA_DB_FILENAME "metadata.db"
+#define LOGS_TABLE "Logs"
+#define BLOBS_TABLE "Blobs"
+
+static uv_loop_t *db_loop;
+static sqlite3 *main_db;
+static char *main_db_dir;  /**< Directory where all the log management databases and log blobs are stored in **/
+static char *main_db_path; /**< Path of MAIN_DB **/
+
+void db_set_lock(uv_mutex_t *db_mut) { 
+	uv_mutex_lock(db_mut); 
+}
+
+void db_release_lock(uv_mutex_t *db_mut) { 
+	uv_mutex_unlock(db_mut);
+}
+
+/**
+ * @brief Throws fatal SQLite3 error
+ * @details In case of a fatal SQLite3 error, the SQLite3 error code will be 
+ * translated to a readable error message and logged to stderr. 
+ * @param[in] rc SQLite3 error code
+ * @param[in] line_no Line number where the error occurred 
+ */
+static inline void fatal_sqlite3_err(int rc, int line_no){
+	debug(D_LOGS_MANAG, "SQL ERROR:%s", sqlite3_errstr(rc));
+	fatal("SQLite error: %s (line %d)", sqlite3_errstr(rc), line_no);
+}
+
+/**
+ * @brief Throws fatal libuv error
+ * @details In case of a fatal libuv error, the libuv error code will be 
+ * translated to a readable error message and logged to stderr. 
+ * @param[in] rc libuv error code
+ * @param[in] line_no Line number where the error occurred 
+ */
+static inline void fatal_libuv_err(int rc, int line_no){
+	fatal("libuv error: %s (line %d)", uv_strerror(rc), line_no);
+}
+
+/**
+ * @brief Get version of SQLite
+ * @return String that contains the SQLite version. Must be freed.
+ */
+char *db_get_sqlite_version() {
+    int rc = 0;
+    sqlite3_stmt *stmt_get_sqlite_version;
+    rc = sqlite3_prepare_v2(main_db,
+                            "SELECT sqlite_version();", -1, &stmt_get_sqlite_version, NULL);
+    if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+    rc = sqlite3_step(stmt_get_sqlite_version);
+    if (unlikely(rc != SQLITE_ROW)) fatal_sqlite3_err(rc, __LINE__);
+    char *text = mallocz(sqlite3_column_bytes(stmt_get_sqlite_version, 0) + 1);
+    strcpy(text, (char *)sqlite3_column_text(stmt_get_sqlite_version, 0));
+    rc = sqlite3_finalize(stmt_get_sqlite_version);
+    if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+    return text;
+}
+
+static void db_writer(void *arg){
```

Comment:
Please, split this function in different functions. I suggest you the following thing:

- In ideal world we would have one function to prepare each one of the prepared statements, but I think it will be easier to have only one function preparing all statements.
- we should have one function to insert data for each prepared statement. And case it fails, we should do what we are doing inside our database. I know that due the fact we are in a transaction, this could remove all previous inserted data, but the data check should happen before we go to the database step.